lib/fluent/plugin/buffer.rb

```diff
@@ -211,17 +223,43 @@ def write(metadata_and_data, bulk: false, enqueue: false)
           operated_chunks.each do |chunk|
             begin
               chunk.commit
-              enqueue_chunk(chunk.metadata) if enqueue || chunk_size_full?(chunk)
+              if enqueue || chunk.unstaged? || chunk_size_full?(chunk)
+                enqueued_chunks << chunk
+              end
               chunk.mon_exit
             rescue => e
               chunk.rollback
               chunk.mon_exit
               errors << e
             end
           end
-          operated_chunks.clear if errors.empty?
 
-          @stage_size += staged_bytesize
+          # All locks about chunks are released.
+
+          synchronize do
+            # At here, staged chunks may be enqueued by other threads.
+            @stage_size += staged_bytesize
+
+            enqueued_chunks.each do |c|
+              if c.staged? && (enqueue || chunk_size_full?(c))
+                m = c.metadata
+                enqueue_chunk(m)
+                if unstaged_chunks[m]
+                  u = unstaged_chunks[m].pop
+                  if u.unstaged? && !chunk_size_full?(u)
```

Comment:
`chunk_size_full?(c)` at L244 is about another chunk `c`, and one at L249 is about `u`.